winrm/winrm.go

```diff
@@ -0,0 +1,271 @@
+}
+
+// NewClient binds new winrm client for initiating connections with the winrm listener
+// Note: use this function before winrm.LoadClientKeys()
+func NewClient(user, host string, gp GetPasswd, stdout io.Writer) *Client {
+	cli := &Client{}
+	winrm.DefaultParameters.TransportDecorator = nil
+
+	var endpoint *winrm.Endpoint
+	if ca := CACert(); ca != nil {
+		endpoint = winrm.NewEndpoint(host, httpPort, false, false, ca, nil, nil, 30*time.Second)
+	} else {
+		endpoint = winrm.NewEndpoint(host, httpPort, false, true, nil, nil, nil, 30*time.Second)
+		logger.Warningf("Skipping CA server verification, using the --insecure option")
+	}
+
+	logger.Debugf("Creating WinRM clinet connection on %s port %d", host, httpPort)
+
+	// if we don't have defined user
+	// use the default administrator on windows machine
+	if user == "" {
+		user = defaultWinndowsUser
+	}
+
+	if gp != nil { // get password from a given way
+		fmt.Fprintf(stdout, "[Winrm] >> ")
+		cli.pass, cli.err = gp()
+		if cli.err != nil {
+			return cli
+		}
+		fmt.Fprintf(stdout, "\n")
+	}
+
+	cli.conn, cli.err = winrm.NewClient(endpoint, user, cli.pass)
+	if cli.err != nil {
+		cli.err = errors.Annotatef(cli.err, "Connot create WinRM client conn")
+	}
+
+	return cli
+}
+
+// Run command and direct output to stdout and errors to stderr
+// If the Run successfully executes the cmd it returns nil on call of Error()
+func (c *Client) Run(command string, stdout io.Writer, stderr io.Writer) {
+	if c.err != nil {
+		return
+	}
+
+	logger.Debugf("Runing cmd on WinRM connection %q", command)
+	_, c.err = c.conn.Run(command, stdout, stderr)
+	if c.err != nil {
+		c.err = errors.Annotatef(c.err, "Can't run WinRM command %q", command)
+	}
+}
+
+// ErrAuth returned if the post request is droped due to invalid credentials
+var (
+	ErrAuth = fmt.Errorf("Unauthorized request")
+	ErrPing = fmt.Errorf("Ping failed, can't recive any response form target")
+)
+
+// Ping checks if we can make a WinRM conn use this client and runs a script
+// if the conn and script successfully runs it returns nil on call of Error()
+// If the default client of this pkg is not set this will panic.
+func (c *Client) Ping() {
+	if c.err != nil {
+		return
+	}
+
+	logger.Debugf("Pinging WinRM listener")
+	var stdout, stderr bytes.Buffer
+
+	c.Run("ECHO HI!", &stdout, &stderr)
+	if c.err != nil {
+		if strings.Contains(c.err.Error(), "401") {
+			logger.Warningf("%s", "The machine blocked due to unathorized access")
+			c.err = ErrAuth
+		} else {
+			logger.Errorf("%s", c.err)
+			c.err = errors.Annotatef(c.err, ErrPing.Error())
+		}
+		return
+	}
+
+	if stderr.Len() != 0 {
+		// save error
+		c.err = fmt.Errorf("%v", strings.TrimSpace(stderr.String()))
+		return
+	}
+
+	logger.Debugf("Output of the ping maches %q", strings.Contains(stdout.String(), "HI!"))
+	if !strings.Contains(stdout.String(), "HI!") {
+		c.err = ErrPing
+		return
+	}
+}
+
+// UseClient set's the default client
+func UseClient(client Exec) {
```

Comment:
What value are the default client & top-level functions providing? I don't want to see them used in Juju, and this code is all about making winrm easier for Juju... so I'm inclined to think that we should drop the default client, UseClient, and other related top-level functions.